### PR TITLE
[#20479] - Skip "From" page on wallet home

### LINKED ISSUE #20479: "From" page is not skipped after long tapping "Send" on the wallet main page
It might be good to fix this together with https://github.com/status-im/status-mobile/issues/20472

### Steps:
1. Go to the wallet main page
2. Long tap the asset -> tap 'send' option

### Actual result:
The FROM page is shown

https://github.com/status-im/status-mobile/assets/52490791/9b8d4975-90d6-441d-ad85-c616decff6b1

### Expected result:

- If "Send" button is long-tapped from Wallet Home page and the selected token or collectible to be sent is available on at least two accounts
  - If the user has Account 1 (with 1 DAI), Account 2 (with 0 DAI), then the FROM page should be skipped.
  - If the user has Account 1 (with 1 DAI), Account 2 (with 1 DAI), and Account 3 (with 0 DAI), then on the FROM page, the user should see only Account 1 and Account 2. Account 3 should not be present or should be disabled.

discussed with the design team: https://discord.com/channels/624634427930312714/1248618318735081544/1252261800716668989

### OS: 
IOS, Android

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

### PR DESCRIPTION
**Pull request by ulisesmac**
fixes #20479

### Summary

This PR is not addressing 
- #20472

yet, but it'll be addressed in a follow-up.

The "From" page now it's not shown when we only own one account with the desired asset to send.

#### Demo 1 (ETH):

https://github.com/user-attachments/assets/940b48a9-0d3c-4cbd-8e66-d7b0a2730f4b

#### Demo 2 (USDC):

https://github.com/user-attachments/assets/e8282b8a-12b1-4c44-a64e-6ccd6e58a742

#### Demo 3 (Collectible):

https://github.com/user-attachments/assets/5bd4060a-4d44-4e83-a24b-7625d24c9d73

### Review notes
Right now, a lot of flows depend on the current state of the re-frame DB, so, in order to not break anything and keep things flexible, some changes have been made in the APIs. These changes are made in a flexible way (by extending the APIs instead of adding constraints to them).

### Testing notes

Please test the existing send flows are still working, I've tested them but, since the mechanism is becoming more and more complex, there might be a forgotten edge case.

### Steps to test

- Open Status
- Own two accounts having the some assets in common
- On the wallet home page, long press on send over the token. The "From" page should be skipped if there's only one account owning the asset.
- All the existing flows should work as expected

status: ready

### CODE REVIEW

File: src/status_im/contexts/wallet/common/token_value/view.cljs [hunk 1]
```diff
@@ -16,14 +16,14 @@
...
-  [send-params]
+  [send-params entry-point]
   {:icon                :i/send
    :accessibility-label :send
    :label               (i18n/label :t/send)
    :on-press            (fn []
                           (rf/dispatch [:hide-bottom-sheet])
                           (rf/dispatch [:wallet/clean-send-data])
-                          (rf/dispatch [:wallet/set-token-to-send send-params]))})
+                          (rf/dispatch [:wallet/set-token-to-send send-params entry-point]))})
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/common/token_value/view.cljs:26]: An `entry-point` parameter has been added, this will define where we are pressing the token and what information we should add or remove while navigating forward and backward respectively.

Currently only exists `:wallet-home` (added when we send from the wallet home)

File: src/status_im/contexts/wallet/events.cljs [hunk 2]
```diff
@@ -65,16 +65,27 @@
...
+                 (not ignore-just-completed-transaction?)
+                 (not just-completed-transaction?))
+            (update :wallet dissoc :current-viewing-account-address)
+
+            entry-point-wallet-home?
+            (update-in [:wallet :ui :send] dissoc :entry-point)
+
+            (and entry-point-wallet-home?
+                 (not just-completed-transaction?))
+            (update :wallet dissoc :current-viewing-account-address))})))
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/events.cljs]: This event was complex and now it's more complex :cry: 

This event will remove entries from re-frame when we navigate back but it depends on trhe `entry-point`.

Currently, `:current-viewing-account` is being used as the sender address, this shouldn't have been made like this, because when we remove `current-viewing-account` from the DB, it also affects the UI.

A larger refactor will be needed if we want to simplify the logic, but it's out of scope of this PR.

CC: @status-im/wallet-mobile-devs

File: src/status_im/contexts/wallet/send/events.cljs [hunk 3]
```diff
@@ -200,25 +200,34 @@
...
+         receiver-networks   (get-in db [:wallet :ui :send :receiver-networks])
+         token-networks-ids  (map :chain-id networks)
+         unsupported-token?  (not-any? (set receiver-networks) token-networks-ids)
+         unique-owner        (when (= (count owners) 1)
+                               (first owners))
+         unique-owner-tokens (get-in db [:wallet :accounts unique-owner :tokens])
+         token-data          (or token
+                                 (when (and token-symbol unique-owner)
+                                   (some #(when (= (:symbol %) token-symbol) %)
+                                         unique-owner-tokens)))]
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:210]: Now in this event, we check if there;s only one token owner, if so, then we add the owner's token data to the send flow

File: src/status_im/contexts/wallet/send/events.cljs [hunk 4]
```diff
@@ -230,11 +239,11 @@
...
  (fn [{:keys [db]} [token]]
    (let [{token-networks :networks
-          token-symbol   :symbol}                  token
+          token-symbol   :symbol} token
          receiver-networks                         (get-in db [:wallet :ui :send :receiver-networks])
-         token-networks-ids                        (mapv #(:chain-id %) token-networks)
-         token-not-supported-in-receiver-networks? (not (some (set receiver-networks)
-                                                              token-networks-ids))]
+         token-networks-ids                        (map :chain-id token-networks)
+         token-not-supported-in-receiver-networks? (not-any? (set receiver-networks)
```
Comment (ulisesmac) [inline src/status_im/contexts/wallet/send/events.cljs:239]: Just a refactor

File: src/status_im/subs/wallet/wallet.cljs [hunk 5]
```diff
@@ -464,22 +465,49 @@
  keep-operable-accounts)
 
 (rf/reg-sub
- :wallet/operable-accounts-without-watched-accounts
+ :wallet/operable-accounts
  :<- [:wallet/accounts-without-watched-accounts]
  keep-operable-accounts)
```
Comment (ulisesmac) [inline src/status_im/subs/wallet/wallet.cljs:471]: Renamed because watched accounts can't be  operable accounts, since we don't own them.
The name now is more simple and short.

CC: @smohamedjavid  wdyt?

File: src/status_im/subs/wallet/wallet.cljs [hunk 6]
```diff
@@ -464,22 +465,49 @@
...
+ :<- [:wallet/operable-accounts]
+ (fn [accounts]
+   (let [positive-balance-in-any-chain? (fn [{:keys [balances-per-chain]}]
+                                          (->> balances-per-chain
+                                               (map (comp :raw-balance val))
+                                               (some pos?)))]
+     (as-> accounts $
+       (group-by :address $)
+       (update-vals $ #(filter positive-balance-in-any-chain? (:tokens (first %))))))))
+
```
Comment (ulisesmac) [inline src/status_im/subs/wallet/wallet.cljs:484]: Added this sub to return a map of `address` -> list of  its tokens with positive balance.

We need it to know if we should skip the from page or not

File: src/status_im/subs/wallet/wallet.cljs [hunk 7]
```diff
@@ -464,22 +466,47 @@
...
+ (fn [[accounts addresses-tokens token-symbol token]]
+   (if-let [asset-symbol (or token-symbol (:symbol token))]
+     (let [addresses-with-asset (as-> addresses-tokens $
+                                  (update-vals $ #(set (map :symbol %)))
+                                  (keep (fn [[address token-symbols]]
+                                          (when (token-symbols asset-symbol) address))
+                                        $)
+                                  (set $))]
+       (filter #(addresses-with-asset (:address %)) accounts))
+     accounts)))
```
Comment (ulisesmac) [inline src/status_im/subs/wallet/wallet.cljs:500]: This function works the same as before but it excludes addresses with zero `raw-balance`. Technically, it means the account doesn't own a token, so this  change is more a fix than a feature

File: src/status_im/subs/wallet/wallet.cljs [hunk 8]
```diff
@@ -464,22 +466,47 @@
...
+
+(rf/reg-sub
+ :wallet/operable-addresses-with-token-symbol
+ :<- [:wallet/operable-addresses-tokens-with-positive-balance]
+ (fn [addresses-tokens [_ token-symbol]]
+   (keep (fn [[address tokens]]
+           (some #(when (= (:symbol %) token-symbol) address)
+                 tokens))
+         addresses-tokens)))
```
Comment (ulisesmac) [inline src/status_im/subs/wallet/wallet.cljs:509]: Just a parameterized sub returning a list of addresses owning a given token-symbol

File: src/status_im/contexts/wallet/send/events.cljs [hunk 9]
```diff
@@ -200,25 +200,35 @@
...
+                                           (dissoc :collectible)
+                                           (assoc :token-not-supported-in-receiver-networks?
+                                                  unsupported-token?)))
+              token-symbol (assoc-in [:wallet :ui :send :token-symbol] token-symbol)
+              token-data   (update-in [:wallet :ui :send]
+                                      #(assoc %
+                                              :token              token-data
+                                              :token-display-name (:symbol token-data)))
+              unique-owner (assoc-in [:wallet :current-viewing-account-address] unique-owner)
+              entry-point  (assoc-in [:wallet :ui :send :entry-point] entry-point))
```
Comment (J-Son89) [inline src/status_im/contexts/wallet/send/events.cljs:224]: Entry point is like the screen the user started the wizard flow on?
Can't say I'm a fan of coupling this ui logic to the wizard logic. 
Is there no other way to handle this? why do we need to do it?

Comment (ulisesmac) [reply src/status_im/contexts/wallet/send/events.cljs:224]: Yes, that's exactly what `:entry-point` does.

We need to know where the flow started because the back button will clean different things according to the `:entry-point`.

E.g.
If we start on wallet home and there's only one account owning an asset, we set the `current-viewing-account` as that account and we remove it when we navigate back.

But if we start the flow from inside the account, we can't remove `current-viewing-account` since this leads to an empty state of the view because the user was watching its account.

As I mentioned in a self-review comment, we shouldn't have used `current-viewing-account` (an UI thing) to determine the sender, but a refactor at this point is hard to perform.

File: src/status_im/contexts/wallet/common/token_value/view.cljs (hunk 1 above)
Comment (briansztamfater) [reply src/status_im/contexts/wallet/common/token_value/view.cljs:26]: Could we use the current screen name as the key instead of using arbitrary keys like `:wallet-home`? If not, can we create constants for these keys to make the code more maintainable? This way, we'll have a single source of truth for the possible values of the `entry-point` parameter.

File: src/status_im/contexts/wallet/send/events.cljs (hunk 9 above)
Comment (briansztamfater) [reply src/status_im/contexts/wallet/send/events.cljs:224]: Would the entry-point logic be necessary if we had a navigation graph in our re-frame database, where we could retrieve the navigation stacks/history? I'm considering whether this could be a valid use case to take into account when refactoring the navigation logic in the future. I recall trying to implement this in the past (#19098), but it introduced extra complexity that wasn't worth adding at the time. However, it might be useful now if I didn't misunderstand the issue.

File: src/status_im/contexts/wallet/common/token_value/view.cljs (hunk 1 above)
Comment (ulisesmac) [reply src/status_im/contexts/wallet/common/token_value/view.cljs:26]: @briansztamfater  Sure!

It's a great suggestion, I'll apply it!

File: src/status_im/contexts/wallet/send/events.cljs (hunk 9 above)
Comment (ulisesmac) [reply src/status_im/contexts/wallet/send/events.cljs:224]: Yes, I think we could skip the entry point if we know the current navigation stack.

This is a point to consider when we change the navigation library, we can store the navigation in re-frame. I've done it in the past and it's not complicated in React Navigation.
-  https://github.com/status-im/status-mobile/issues/20471

**Comment by VolodLytvynenko** [comment]
hi @ulisesmac Thank you for fixes. Apologies, I was not so familiar with the flow when I reported the [original issue](https://github.com/status-im/status-mobile/issues/20479). We have some more places from which we can be navigated to the "FROM" page.

Would it be better to fix this separately or in the current PR?

## ISSUE 1: The FROM page is shown when user goes to send flow using scanner

### Precondition:
User has only one account

### Steps:
1. Goes to scanner -> scan the wallet QR

### Actual result:
User sees the 1 account in from page

https://github.com/user-attachments/assets/66f06ead-a7b1-436f-aec0-54683201dc1b


### Expected result:
User is navigated directly to the send flow page if user has only 1 account and use scanner during send flow

**Comment by VolodLytvynenko** [comment]
## ISSUE 2: From is shown when user goes to sending flow via bridge

### Preconditions:
User has Account 1 

### Steps:
1. long tap the asset
3. Tap "bridge"

### Actual result:
User sees the FROM page

https://github.com/user-attachments/assets/0b130e15-3e7d-4c5c-8368-c5095ad00170


### Expected result:
when "bridge" button is long-tapped from Wallet Home page and the selected token to be sent is available on at least two accounts:
- If the user has Account 1 (with 1 DAI), Account 2 (with 0 DAI), then the FROM page should be skipped.
- If the user has Account 1 (with 1 DAI), Account 2 (with 1 DAI), and Account 3 (with 0 DAI), then on the FROM page, the user should see only Account 1 and Account 2. Account 3 should not be present or should be disabled.

**Comment by ulisesmac** [comment]
@VolodLytvynenko Thanks for reporting these entry points.

I'll fix them in this PR.

## UPDATE:

### Issue 1
I think this issue is completely different, it's not related to skip the "From" page when only one user owns an asset, it could be written as: "Skip From page when there's only one account in the device", so this should be addressed separately and the issue should already have the answers for the following quesitons:

If there's no assets and there's only one account, should we completely remove the "send" flow from everywhere (not only the "From" page)? How would they look? what will we do instead? Do we need need new designs/states for these cases?

### Issue 2
The bridge flow is a completely different flow compared to send, it requires a deeper understanding of the data it handles and its flows/entry-points, so it'd signifincantly require more time to get solved and this PR will grow a lot in terms of code. Better to solve it separately.

What I did that works well is omiting the Bridge option when there is zero accounts owning an asset:

https://github.com/user-attachments/assets/6b0670b9-6216-4507-ac9c-4f9cae897848

The problem is not just removing the From page, it's about making sure the other flows still show the option and nothing is broken, so, again, better to solve it separately.

File: src/status_im/contexts/wallet/common/token_value/view.cljs (hunk 1 above)
Comment (ulisesmac) [reply src/status_im/contexts/wallet/common/token_value/view.cljs:26]: @briansztamfater The wallet home component is not mounted on the screen navigation hierarchy, so that it doesn't have a screen name. Creating the keyword `:screen/wallet-home` for this purpose is misleading, so better to keep it as is, but for other valid entry-points I'll apply your suggestion.

**Comment by ulisesmac** [comment]
@J-Son89 @briansztamfater 
Could you please check the answers and approve the PR if possible?

**Comment by VolodLytvynenko** [comment]
> @VolodLytvynenko Thanks for reporting these entry points.
> 
> I'll fix them in this PR.
> 
> ## UPDATE:
> ### Issue 1
> I think this issue is completely different, it's not related to skip the "From" page when only one user owns an asset, it could be written as: "Skip From page when there's only one account in the device", so this should be addressed separately and the issue should already have the answers for the following quesitons:
> 
> If there's no assets and there's only one account, should we completely remove the "send" flow from everywhere (not only the "From" page)? How would they look? what will we do instead? Do we need need new designs/states for these cases?

hi @ulisesmac Yes, you are right. The expected result should be "Skip the From page when there's only one account on the device." I discussed this with @xAlisher and described it in the expected behavior of [issue 1](https://github.com/status-im/status-mobile/pull/20773#issuecomment-2238930780).

### Requirement 1:
- Case 1 ([original issue](https://github.com/status-im/status-mobile/issues/20479)): When an asset is long-pressed and the "send" or "bridge" button is tapped, check if the same asset is present on other accounts. If yes, skip the FROM page.
- Case 2: When sending begins via the scanner, if more than one account exists (excluding watch-only accounts), show the FROM page. Otherwise, skip it.

### Requirement 1 (alternative implementation):
If Requirement 1 is too complicated to implement:

For both sending via the scanner and when an asset is long-pressed and the "send" or "bridge" button is tapped, if more than one account exists (excluding watch-only accounts), show the FROM page. Otherwise, skip it.

File: src/status_im/contexts/wallet/common/token_value/view.cljs (hunk 1 above)
Comment (briansztamfater) [reply src/status_im/contexts/wallet/common/token_value/view.cljs:26]: AFAIK, you can get the screen name from `:view-id` key in app-db. From what I could find by debugging with re-frisk, "wallet home" screen name is `:wallet-stack` (as it is the root screen of wallet tab stack).

**Review by briansztamfater — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
hi @ulisesmac thank you for PR. no issues from my side. PR is ready to be merged

**Review by J-Son89 — APPROVED** [review]
